### PR TITLE
Export plugin (deprecates/replaces "Export Plot")

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request implements the basic UI functionality of a future "Export" plugin, to replace the "Export Plot" plugin.  ~This PR should not yet be merged until functionality from the "Export Plot" plugin is mapped over and properly deprecated, but can be merged before additional support is added as they are hidden under feature flags.~

Without any feature flags enabled (viewer export only - note that cubeviz movie options should also be added underneath the format when functionality from export plot is mapped over):

https://github.com/spacetelescope/jdaviz/assets/877591/d0243f0b-e563-4f5e-ac47-05e4652d960e

To enable feature flags:

```
exp = viz.plugins['Export']
exp._obj.dev_multi_support = True
exp._obj.dev_dataset_support = True
exp._obj.dev_subset_support = True
exp._obj.dev_table_support = True
exp._obj.dev_plot_support = True
exp = viz.plugins['Export'] # to pickup the new user API
exp.open_in_tray()
```

https://github.com/spacetelescope/jdaviz/assets/877591/b0899248-6a41-42ca-924e-3b1df264e198

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by Jenneh** [comment]
I am concerned about this paradigm. Most export options are accessible from the thing (plot, table, image, etc) that is being exported. I worry that users will not innately know to look in this menu for export options. It would feel more natural to have an export icon right on each panel.

Perhaps there could be a button on each plot that opens and expands this panel?

Another note, There were some requests for the ability to export plots from within other plugins. It does not seem like this supports that functionality but that would be desirable.

**Comment by gibsongreen** [comment]
One idea I can think of is, let's say the user extracts a spectrum. In the Export plugin, in the Data Selection, this would auto populate as it should. The extracted spectrum could be the default selection made by the plugin, since this is new data comparatively to the ERR or SCI data which they had from the original file loaded.

Another idea is, let's say the user extracts a spectrum and creates a moment map (or any combination of multiple pieces of new data). Instead of toggling the multi-select button on, if the default idea previously is desirable, the default in this case would use multi-select, and default to the selection of both the spectrum and the moment map for download.

**Comment by gibsongreen** [comment]
The zip idea was mentioned in parking lot. I can imagine that if a user has several pieces of data the desire to export, that renaming each file would be painful. I wasn't sure if I misinterpreted, but I thought the default option is multiple items were selected, the renaming of the zip file would be what the filename input field would allow. 

Whenever I've downloaded compressed folders, I am normally not thinking about the folder name, but rather the files inside. I do think giving the option to rename each data selection would be useful. 

This being said, I do like the layout and the way each piece of data that possible for download is displayed. I don't know if this is possible, but the slow double-click on Mac or Windows for file/directory renaming would be awesome here. The current display could be kept in this case, and the user could selectively chose which files they wanted to rename without a bunch of input text-boxes.

**Comment by gibsongreen** [comment]
> One idea I can think of is, let's say the user extracts a spectrum. In the Export plugin, in the Data Selection, this would auto populate as it should. The extracted spectrum could be the default selection made by the plugin, since this is new data comparatively to the ERR or SCI data which they had from the original file loaded.
> 
> Another idea is, let's say the user extracts a spectrum and creates a moment map (or any combination of multiple pieces of new data). Instead of toggling the multi-select button on, if the default idea previously is desirable, the default in this case would use multi-select, and default to the selection of both the spectrum and the moment map for download.

@Jenneh idea for the button at the end of the plugins for exporting. If the user had generated data that they wanted to export, and then clicked this button, the default irregardless could have this data pre-selected in the Export plugin. More importantly, when the this button is clicked, the Export plot could open, and the data generated could be temporarily highlighted in the Export Plugin, to inform them of where their desired export data is within the lists of selections.

**Comment by pllim** [comment]
FWIW, this is how Ginga does it. I implemented it for TEL based on their functional requirements:

https://ginga.readthedocs.io/en/stable/manual/plugins_global/saveimage.html

Which is used to extend it to TEL specific functionality here:

https://wss-tools.readthedocs.io/en/latest/wss_tools/quip/savequip.html

Not the best design but in case you find it useful as a reference.

**Comment by kecnry** [comment]
This is now ready for review as a replacement of the previous "Export Plot" plugin, aiming to provide all the same functionality as previously, but with the design shown above and hidden behind the feature flags to move forward.

Regarding the design review comments above, we are planning to improve the shortcut icon in the viewers and add the capability to directly scroll to the plugin when clicked.  We can also consider buttons in places where creating an instance of this plugin, prepopulated, in a popout or overlay would be beneficial.

**Comment by kecnry** [comment]
Re the low patch coverage: most of these lines are either under the feature-flags or were not covered in the previous implementation either.

**Comment by rosteen** [comment]
Previously, if you didn't have `opencv-python` installed, there would be a warning under the export video section saying it needed to be installed to export to video. Now there's nothing, the mp4 option is just missing from the format menu in that case. I think we need something in the UI to tell the user that video export would be available if they installed that package.

**Comment by kecnry** [comment]
Ok, let me try to code that up - the downside is that it will require extra checks on the python side as well and won't be as clean from the API, but we can compare the two options and decide.

**Comment by kecnry** [comment]
ok, now mp4 is excluded from the list for non-cubeviz or when running with an external server (i.e. MAST), but is included in the list otherwise.  If its included and OPENCV is not installed, then an alert warning replaces the options and calling export from the API will raise an error.

**Comment by rosteen** [comment]
Is the`format` field planned to apply to the plots and tables as well? Currently it seems a little confusing that it says "Image format for exporting viewers" - I don't know what file format I would get if I exported a plot or table. Maybe the format menu could beat the bottom, below the other selections, and populate based on what you picked to export?

**Comment by kecnry** [comment]
Each section will likely need its own format dropdown, so that batch exporting can select a different format per section (png for viewers, fits for data, etc), and to avoid having to change the options and revert previous selections, etc.

**Review by cshanahan1 — COMMENTED** [review]
if you are saving from a viewer and re-save with the same filename, it will overwrite the file (maybe a warning or some indication of this in the plugin so it's clear?). 

however, if you keep the same filename and switch to another viewer to save, it will append a number to the filename so that it doesn't overwrite. this makes sense to me. 

however, once you make the switch to a new viewer, and save the file again and again with the same name, it starts incrementing the filename every time rather than overwriting, which is different behavior than you get when you start on one of the viewers. for consistency, if you are on one viewer and are re-saving, it should either overwrite or not but right now it is doing both depending on if you had toggled at one point.

**Comment by kecnry** [comment]
I don't believe any of that behavior was introduced here and I'm trying to keep the scope limited to reproducing previous functionality but in the new plugin.  I agree that that is quite confusing though, so please do open issues/tickets so we can improve on the behavior.

**Comment by pllim** [comment]
Re: naming -- Maybe we should start appending timestamp like the OS screenshot apps.

**Comment by cshanahan1** [comment]
Do we want the output of model fitting to be something that can be saved?

**Comment by cshanahan1** [comment]
(sorry for spam, im just commenting as I go through this).

The moment map and spectral extraction plugins have a 'save to fits' button. Will those eventually be removed and moved to the new export plugin?

**Comment by kecnry** [comment]
> Do we want the output of model fitting to be something that can be saved?

That will fall under both the plugin tables or data sections, once those are implemented.

> The moment map and spectral extraction plugins have a 'save to fits' button. Will those eventually be removed and moved to the new export plugin?

I think so, yes, we could have a shortcut from those plugins that populate the sections here, or popout and instance of this plugin separate from the one in the tray.  Again, that requires export data functionality which is not included in this PR.

**Comment by cshanahan1** [comment]
thanks for clarifying. i was assuming this PR was adding all 'savable' things to the export plugin (like subsets) without the ability to actually save them yet.

**Review by rosteen — APPROVED** [review]
LGTM, all my concerns were either already existing in `main` or things that will be addressed as we build out the functionality of the currently hidden stuff.

File: docs/cubeviz/plugins.rst [hunk 1]
```diff
@@ -326,8 +326,8 @@ have valid flux units. For 3D data, the current :ref:`slice` is used.
 
 .. _cubeviz-export-plot:
```
Comment (cshanahan1) [inline docs/cubeviz/plugins.rst:327]: should this reference name be changed from _cubeviz-export-plot to cubeviz-export (in this and in {otherconfigs)/plugins.rst?)

Comment (kecnry) [reply docs/cubeviz/plugins.rst:327]: I wasn't sure, I didn't want to break existing RTD links so left them alone, but maybe we should put in redirects.... @pllim - what's the best path forward here?

Comment (pllim) [reply docs/cubeviz/plugins.rst:327]: The reference name isn't user facing, I would just leave it so existing intersphinx won't break.

**Comment by cshanahan1** [comment]
theres a small reference to the old plugin name in test_template_mixin on line 113

**Comment by cshanahan1** [comment]
Otherwise, my concerns are all out of scope so i will approve. 

(One more out-of-scope just so I don't forget - right now when you select a viewer to export, a dropdown to select file type appears. This will happen for the other items once the export is actually implemented, correct?)

**Review by cshanahan1 — APPROVED** [review]

**Comment by kecnry** [comment]
> One more out-of-scope just so I don't forget - right now when you select a viewer to export, a dropdown to select file type appears. This will happen for the other items once the export is actually implemented, correct?

Yes, we'll need to implement per-"section" format dropdowns, as needed.

Please make sure all of your out-of-scope comments are covered by issues/tickets!

**Comment by kecnry** [comment]
(rebasing will be a bit complicated - I'll work on that and fix the reference before merging)

**Pull request merged**